web/authentication.go

```diff
@@ -0,0 +1,291 @@
+func (x *X509Authenticator) verifyPeerCertificate(cert *x509.Certificate, _ [][]*x509.Certificate) error {
+	// Build up a slice of strings with all Subject Alternate Name values
+	sanValues := append(cert.DNSNames, cert.EmailAddresses...)
+
+	for _, ip := range cert.IPAddresses {
+		sanValues = append(sanValues, ip.String())
+	}
+
+	for _, uri := range cert.URIs {
+		sanValues = append(sanValues, uri.String())
+	}
+
+	for _, sanValue := range sanValues {
+		for _, allowedSan := range x.clientAllowedSans {
+			if sanValue == allowedSan {
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("could not find allowed SANs in client cert, found: %v", x.clientAllowedSans)
+}
+
+func NewX509Authenticator(clientAuth tls.ClientAuthType, clientCAs *x509.CertPool, clientAllowedSans []string) Authenticator {
+	return &X509Authenticator{
+		clientAuth:        clientAuth,
+		clientCAs:         clientCAs,
+		clientAllowedSans: clientAllowedSans,
+	}
+}
+
+var _ Authenticator = &X509Authenticator{}
+
+type BasicAuthAuthenticator struct {
+	users map[string]config.Secret
+
+	cache *cache
+	// bcryptMtx is there to ensure that bcrypt.CompareHashAndPassword is run
+	// only once in parallel as this is CPU intensive.
+	bcryptMtx sync.Mutex
+}
+
+func (b *BasicAuthAuthenticator) Authenticate(r *http.Request) (bool, string, error) {
+	user, pass, auth := r.BasicAuth()
+
+	if !auth {
+		return false, "No credentials in request", nil
+	}
+
+	hashedPassword, validUser := b.users[user]
+
+	if !validUser {
+		// The user is not found. Use a fixed password hash to
+		// prevent user enumeration by timing requests.
+		// This is a bcrypt-hashed version of "fakepassword".
+		hashedPassword = "$2y$10$QOauhQNbBCuQDKes6eFzPeMqBSjb7Mr5DUmpZ/VcEd00UAV/LDeSi"
+	}
+
+	cacheKey := strings.Join(
+		[]string{
+			hex.EncodeToString([]byte(user)),
+			hex.EncodeToString([]byte(hashedPassword)),
+			hex.EncodeToString([]byte(pass)),
+		}, ":")
+	authOk, ok := b.cache.get(cacheKey)
+
+	if !ok {
+		// This user, hashedPassword, password is not cached.
+		b.bcryptMtx.Lock()
+		err := bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(pass))
+		b.bcryptMtx.Unlock()
+
+		authOk = validUser && err == nil
+		b.cache.set(cacheKey, authOk)
+	}
+
+	if authOk && validUser {
+		return true, "", nil
+	}
+
+	return false, "Invalid credentials", nil
+}
+
+func NewBasicAuthAuthenticator(users map[string]config.Secret) Authenticator {
+	return &BasicAuthAuthenticator{
+		cache: newCache(),
+		users: users,
+	}
+}
+
+var _ Authenticator = &BasicAuthAuthenticator{}
+
+type AuthenticationExceptor interface {
+	IsExcepted(r *http.Request) bool
+}
+
+type AuthenticationExceptorFunc func(*http.Request) bool
+
+func (f AuthenticationExceptorFunc) IsExcepted(r *http.Request) bool {
+	return f(r)
+}
+
+func WithAuthenticationExceptor(handler http.Handler, authenticator Authenticator, exceptor AuthenticationExceptor, logger log.Logger) http.Handler {
```

Comment:
e.g. at the mux level you already have a way to specify which path go to which handles and you can make richer rules then inside another layer of exact path matching

Reply:
It would definitely be a nicer approach, but would you be able to handle changes to the config file in that way? It's supported with the current implementation (albeit very poorly - by reading the config file on every request... multiple times) so I suppose it's expected to support it.

Reply:
> but would you be able to handle changes to the config file in that way

Is this asking about dynamically changing which paths require auth? I don't expect that to by dynamic at all. (There is a fixes set of unauthorized endpoints like /readyz and then the rest that can either have auth enabled or disabled.)

Reply:
> I don't expect that to by dynamic at all.

Yeah, but currently exporter-toolkit "reloads" the config file for all configuration. I suppose it has to be consistent for the excluded paths as well.

> There is a fixes set of unauthorized endpoints like /readyz and then the rest that can either have auth enabled or disabled.

This (maybe) applies to prometheus, but it won't necessarily be the case for the entire toolkit.

Reply:
> I suppose it has to be consistent for the excluded paths as well.

I was proposing not to put the paths into the config at all but rather into the code

Reply:
I agree that it's a better way to do it, but it looks like this is what the prometheus maintainers were aiming for: https://github.com/prometheus/prometheus/issues/9166#issuecomment-896173714